src/MSMToolsCluster.jl

```diff
@@ -0,0 +1,259 @@
+    end
+
+    closed_components_ev_flat = vcat(closed_components_ev...)
+    closed_components_enum_flat = vcat(closed_components_enum...)
+
+    # cluster each component
+    component_indexes = closed_components_enum_flat[sortperm(closed_components_ev_flat)][1:m]
+    ipcca = 1
+    for i in 1:n_closed_components
+        component = closed_components[i]
+        m_by_component = count(x -> x == i, component_indexes) # number of PCCA states per component
+
+        if m_by_component == 1 # trivial case
+            chi[component, ipcca] .= 1.0
+            ipcca += 1
+        elseif m_by_component > 1
+            chi[component, ipcca:(ipcca + m_by_component - 1)] .= pcca_connected(
+                closed_components_Psub[i], m_by_component, π == nothing ? nothing : π[component]
+            )
+            ipcca += m_by_component
+        else
+            throw(error("Did not expect component $i to have $m_by_component."))
+        end
+    end
+
+    # assign transition states
+    if !isempty(transition_states)
+        # make all closed states absorbing, so we can see which closed state we hit first
+        Pabs = copy(P)
+        Pabs[closed_states, :] .= 0.0
+        Pabs[closed_states, closed_states] .= 1.0
+        for i in closed_states
+            h = hitting_probability(Pabs, i) # hitting probability to each closed state
+            for j in transition_states
+                chi[j, :] .+= h[j] .* chi[i, :]
+            end
+        end
+    end
+
+    nmeta = count(x -> x != 0, sum(chi, dims=1))
+    if nmeta < m
+        throw(error("Requested $m metastable states, but transition matrix only has $nmeta. " *
+                "Try requesting fewer metastable states."))
+    end
+
+    return chi
+end
+
+# Reimplementation of PCCA+ spectral clustering method with optimized memberships (assumes fully connected transition matrix)
+# https://github.com/deeptime-ml/deeptime/blob/a6ac0b93a55d688fe8f3af119680105763366220/deeptime/markov/tools/analysis/dense/_pcca.py#L197
+function pcca_connected(P, n, pi=nothing)
+    labels = connected_sets(P)
+    n_components = length(labels)
+    if n_components > 1
+        throw(error("Unable to use pcca_connected, transition matrix is not fully connected."))
+    end
+
+    if pi === nothing
+        pi = stationary_distribution(P)
+    else
+        if size(pi, 1) != size(P, 1)
+            throw(error("Stationary distribution must span the entire state space but got $(size(pi, 1)) states " *
+                                "instead of $(size(P, 1))."))
+        end
+        pi ./= sum(pi)  # normalization
+    end
+
+    eigenvalues, eigenvectors = eigen(P)
+    indices = sortperm(eigenvalues, rev=true)
+    evecs = eigenvectors[:, indices]
+
+    # orthonormalize
+    for i in 1:n
+        evecs[:, i] /= sqrt(dot(evecs[:, i] .* pi, evecs[:, i]))
+    end
+
+    # make 1st eigenvector positive
+    evecs[:, 1] = abs.(evecs[:, 1])
+
+    if !all(isreal.(evecs))
```

Comment:
For operations like `!all(...)` it's often more convenient to use the function `any`, which returns true as soon as the condition is met, which can be a negative conditon

Reply:
fixed in e7d54264dc57427fc1287be2e5b79708c1a2fa2e